### PR TITLE
Make AST->HIR lowering incremental

### PR DESCRIPTION
**Pull request by cjgillot**
Implementation of MCP https://github.com/rust-lang/compiler-team/issues/452

AST->HIR lowering is currently done ahead of the construction of the incremental compilation engine. This issue proposes to make lowering incremental by introducing a query `lower_to_hir(def: LocalDefId) -> Option<&hir::OwnerInfo<'_>>` which lowers the AST item corresponding to definition def into the corresponding HIR and some associated information.

The objective is to eventually skip lowering some parts of the AST, and to allow progress towards making resolution and expansion incremental.

The proposed design is as follows:

* remove global analyses from the current lowering pass, and pull them back to resolution stage;
* attach lowering information to definitions instead of having multiple maps inside `hir::Crate`, effectively making `hir::Crate` an `IndexVec<LocalDefId, Option<OwnerInfo<'_>>>`;
* extract the required resolution outputs from the resolver and keep them immutable;
* allow to create new definitions from inside the query system (*);
* index the AST by `LocalDefId`, using an `IndexVec<LocalDefId, AstOwner>` where `AstOwner` is an enum holding `OwningRef<Lrc<Crate>, ItemLike>` for `ItemLike` different AST item types;
* perform HIR indexing per-owner;
* make lowering incremental.

One of the most subtle changes is (*), which allows to create new definitions from inside queries. It is made safe using an eval-always query which forces re-execution of its caller, and effectively enforces the side-effects creating the definition. The first time definitions are iterated over (for instance for metadata output), a read-lock to the definitions is leaked, which seals the list of definitions.

This PR is blocked on:
- [x]  rust-lang/rust#87234;
- [x]  rust-lang/rust#88019;
- [x]  rust-lang/rust#88703;
- [x]  rust-lang/rust#88880;
- [x]  rust-lang/rust#89124;
- [x]  rust-lang/rust#90146;
- [x]  rust-lang/rust#91112;
- [x]  rust-lang/rust#89278;
- [x]  rust-lang/rust#90451;
- [x]  rust-lang/rust#90204;
- [x] rust-lang/rust#95573.

The query scheme should basically look like this:
- the untracked ast is used to compute lower_to_hir;
- lower_to_hir yields hir_attrs, hir_owner_nodes and hir_owner_parent;
- hir_owner_nodes yields hir_owner;
- hir_module_items uses hir_owner_nodes to gather all the items inside a module;
- hir_crate uses hir_owner_nodes to gather all the items inside the crate.

r? @michaelwoerister as requested on Zulip

### CODE REVIEW

File: compiler/rustc_query_impl/src/on_disk_cache.rs [hunk 1]
```diff
@@ -794,6 +793,14 @@ impl<'a, 'tcx> Decodable<CacheDecoder<'a, 'tcx>> for &'tcx [Span] {
     }
 }
 
+impl<'tcx, 'a> Decodable<CacheDecoder<'a, 'tcx>> for &'tcx [LocalDefId] {
+    #[inline]
+    fn decode(d: &mut CacheDecoder<'a, 'tcx>) -> Self {
+        let ids: std::borrow::Cow<'_, [LocalDefId]> = Decodable::decode(d);
```
Comment (bjorn3) [inline compiler/rustc_query_impl/src/on_disk_cache.rs:799]: Why does this go through `Cow`?

**Comment by cjgillot** [comment]
This PR features a sizeable perf regression, up to 11%. I believe this is entirely due to query overhead: this PR requires one extra query invocation for each `LocalDefId` in the crate. I split https://github.com/rust-lang/rust/pull/95573 which makes lowering a query, but does not separate lowering each `LocalDefId`.

**Comment by JohnCSimon** [comment]
@cjgillot 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**